sa/model_test.go

```diff
@@ -224,3 +239,144 @@ func TestPopulateAttemptedFieldsBadJSON(t *testing.T) {
 		})
 	}
 }
+
+func TestCerficatesTableContainsDuplicateSerials(t *testing.T) {
+	testCtx := setup(t)
+	defer testCtx.cleanUp()
+
+	serialString := core.SerialToString(big.NewInt(1337))
+
+	// Register and insert a certificate with a serial of `1337`.
+	testCtx.setupCertificate1337A(t)
+
+	// This should return the certificate that we just inserted.
+	_, err := SelectCertificate(testCtx.db, serialString)
+	test.AssertNotError(t, err, "received an error for a valid query")
+
+	// Register and insert a certificate with a serial of `1337` but for a
+	// different hostname.
+	testCtx.setupCertificate1337B(t)
+
+	// With a duplicate being present, this should error.
+	_, err = SelectCertificate(testCtx.db, serialString)
+	test.AssertError(t, err, "should've received an error for multiple rows")
+}
+
+type testCtx struct {
+	db      *db.WrappedMap
+	ssa     core.StorageAdder
+	cleanUp func()
+}
+
+func (c testCtx) registerCertificate(t *testing.T, emailRaw string, jsonKey []byte, regID int64) {
```

Comment:
This should be `registerAccount` or `createAccount` or similar. Also, is this needed? Does SelectCertificate fail if the account doesn't exist?

Reply:
Ah yeah, fair point. I think at one point I was operating under the assumption that there were key constraints between the `certificates` and `registrations` tables.